contractcourt/chain_watcher_test.go

```diff
@@ -213,3 +223,154 @@ func TestChainWatcherRemoteUnilateralClosePendingCommit(t *testing.T) {
+	// commitment that is beyond our best known commitment for them, and
+	// they don't have a pending commitment (one we sent but which hasn't
+	// been revoked), then we'll properly detect this case, and execute the
+	// DLP protocol on our end.
+	//
+	// broadcastStateNum is the number that we'll trick Alice into thinking
+	// was broadcast, while numUpdates is the actual number of updates
+	// we'll execute. Both of these will be random 8-bit values generated
+	// by testing/quick.
+	dlpScenario := func(broadcastStateNum, numUpdates uint8) bool {
+		// If the random state num is less than our current state num,
+		// then this case will pass as we only care about the DLP case.
+		if broadcastStateNum <= numUpdates {
+			return true
+		}
+
+		// First, we'll create two channels which already have
+		// established a commitment contract between themselves.
+		aliceChannel, bobChannel, cleanUp, err := lnwallet.CreateTestChannels()
+		if err != nil {
+			t.Fatalf("unable to create test channels: %v", err)
+		}
+		defer cleanUp()
+
+		// Within the test, we'll proxy the extractStateNumHint closure
+		// call into this channel so we can feed in arbitrary state
+		// deviations at all, without doing any state updates.
+		stateNumChan := make(chan uint64, 1)
+
+		// With the channels created, we'll now create a chain watcher
+		// instance which will be watching for any closes of Alice's
+		// channel.
+		aliceNotifier := &mockNotifier{
+			spendChan: make(chan *chainntnfs.SpendDetail),
+		}
+		aliceChainWatcher, err := newChainWatcher(chainWatcherConfig{
+			chanState: aliceChannel.State(),
+			notifier:  aliceNotifier,
+			signer:    aliceChannel.Signer,
+			extractStateNumHint: func(*wire.MsgTx,
+				[lnwallet.StateHintSize]byte) uint64 {
+
+				return <-stateNumChan
+			},
+		})
+		if err != nil {
+			t.Fatalf("unable to create chain watcher: %v", err)
+		}
+		if err := aliceChainWatcher.Start(); err != nil {
+			t.Fatalf("unable to start chain watcher: %v", err)
+		}
+		defer aliceChainWatcher.Stop()
+
+		// Based on the number of random updates for this state, make a
+		// new HTLC to add to the commitment, and then lock in a state
+		// transition.
+		const htlcAmt = 1000
+		for i := 0; i < int(numUpdates); i++ {
+			addFakeHTLC(
+				t, 1000, uint64(i), aliceChannel, bobChannel,
+			)
+
+			err := lnwallet.ForceStateTransition(
+				aliceChannel, bobChannel,
+			)
+			if err != nil {
+				t.Errorf("unable to trigger state "+
+					"transition: %v", err)
+				return false
+			}
+		}
+
+		// We'll request a new channel event subscription from Alice's
+		// chain watcher so we can be notified of our fake close below.
+		chanEvents := aliceChainWatcher.SubscribeChannelEvents()
+
+		// Now that we have our updates in place, we'll feed Alice a
+		// random state number. If this is beyond her best known state
+		// for Bob, then she should enter the DLP protocol.
+		stateNumChan <- uint64(broadcastStateNum)
```

Comment:
Why does this have to be sent on a channel? Can't we just make `extractStateNumHint` return `broadcastStateNum`?